File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        HashMap<Integer, List<Integer>> map = new HashMap<>();
        
        for(int i = 0; i < nums.length; ++i){
            if(map.containsKey(nums[i])){
                for(int idx : map.get(nums[i])){
                    if(Math.abs(i - idx) <= k) return true;
                }
            }
            map.putIfAbsent(nums[i], new ArrayList<>());
            map.get(nums[i]).add(i);
        }
        
        return false;
    }
}

File: 0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        n_map = {}
        
        for i in range(len(nums)):
            n = nums[i]
            if n in n_map and i - n_map[n] <= k:
                return True
            
            n_map[n] = i
        return False